node/src/components/network.rs

```diff
@@ -1346,3 +1369,290 @@ where
             .finish()
     }
 }
+
+#[cfg(test)]
+mod gossip_target_tests {
+    use std::{collections::BTreeSet, iter};
+
+    use static_assertions::const_assert;
+
+    use casper_types::testing::TestRng;
+
+    use super::*;
+
+    const VALIDATOR_COUNT: usize = 10;
+    const NON_VALIDATOR_COUNT: usize = 20;
+    // The tests assume that we have fewer validators than non-validators.
+    const_assert!(VALIDATOR_COUNT < NON_VALIDATOR_COUNT);
+
+    struct Fixture {
+        validators: BTreeSet<NodeId>,
+        non_validators: BTreeSet<NodeId>,
+        all_peers: Vec<NodeId>,
+    }
+
+    impl Fixture {
+        fn new(rng: &mut TestRng) -> Self {
+            let validators: BTreeSet<NodeId> = iter::repeat_with(|| NodeId::random(rng))
+                .take(VALIDATOR_COUNT)
+                .collect();
+            let non_validators: BTreeSet<NodeId> = iter::repeat_with(|| NodeId::random(rng))
+                .take(NON_VALIDATOR_COUNT)
+                .collect();
+
+            let mut all_peers: Vec<NodeId> = validators
+                .iter()
+                .copied()
+                .chain(non_validators.iter().copied())
+                .collect();
+            all_peers.shuffle(rng);
+
+            Fixture {
+                validators,
+                non_validators,
+                all_peers,
+            }
+        }
+
+        fn is_validator_in_era(&self) -> impl Fn(EraId, &NodeId) -> bool + '_ {
+            move |_era_id: EraId, node_id: &NodeId| self.validators.contains(node_id)
+        }
+
+        fn num_validators<'a>(&self, input: impl Iterator<Item = &'a NodeId>) -> usize {
+            input
+                .filter(move |&node_id| self.validators.contains(node_id))
+                .count()
+        }
+
+        fn num_non_validators<'a>(&self, input: impl Iterator<Item = &'a NodeId>) -> usize {
+            input
+                .filter(move |&node_id| self.non_validators.contains(node_id))
+                .count()
+        }
+    }
+
+    #[test]
+    fn should_choose_mixed() {
```

Comment:
I'd consider adding one test to make sure we grab a correct number of peers from a set (validators or non-validators).

Currently, we may change the parameter in [this line](https://github.com/casper-network/casper-node/pull/3604/commits/d7b97474f66dd168bd10bb1dce14e593df15f405#diff-ced0e709fad62f0c013bb3d691fb1082aface83249abccb628224b31e62bec65R1096) to `count - 1` and the incorrect behavior will get hidden by `interleave()`, which is "saturating".

Here's how such test may look like:

```
        // Choose 3 from all peers, exclude all non-validators, should return 3 validators,
        let exclude: HashSet<_> = fixture.non_validators.iter().copied().collect();
        let chosen = choose_gossip_peers(
            &mut rng,
            TARGET,
            3,
            exclude,
            fixture.all_peers.iter().copied(),
            fixture.is_validator_in_era(),
        );
        assert_eq!(chosen.len(), 3);
        assert_eq!(fixture.num_validators(chosen.iter()), 3);
```

It should fail 50% of the time if we happen mess-up the `choose_multiple` in the future :)